### PR TITLE
Level2 Tutorial Material

### PR DESCRIPTION
**Pull request by jleidel**
Adding Level2 tutorial material.  Topics covered: 

- Parallel SST Runs
- Advanced python scripting
- Advanced JSON Scripting
- Loading Parallel Graphs
- Links to external SST components

### CODE REVIEW

File: Tutorials/Level1Tutorial.md [hunk 1]
```diff
@@ -0,0 +1,793 @@
...
+  simulation of large machines at multiple levels. The SST has been used in 
+  variety of network designs, memory hierarchies, and applications.
+
+## Components of SST: <a name="ComponentsOfSST"></a>
+  To meet these requirements, the SST is comprised of a simple simulation core 
+  that contains a parallel discrete event simulator (DES) and support services 
+  for simulation. components, representing hardware systems such as processors, 
+  network switches, or memory devices, interface with the simulation core to 
+  communicate and operate with a common notion of timeframe. The simulation 
+  core also provides support services such as power and area estimation,
```
Comment (gvoskuilen) [inline Tutorials/Level1Tutorial.md]: No support for power/area estimation or checkpointing.

File: Tutorials/Level1Tutorial.md [hunk 2]
```diff
@@ -0,0 +1,793 @@
...
+  communicate and operate with a common notion of timeframe. The simulation 
+  core also provides support services such as power and area estimation, 
+  checkpointing, configuration/initialization of the simulation, and statistics 
+  gathering. The SST’s modular interface eases the integration of existing 
+  simulators into a common framework and is licensed under a BSD-like license. 
+  In addition, the components of the SST model are compatible with external 
+  models such as Gem5 and DRAMSim2. Because of the open source core, the 
+  modular framework is easily extensible with new models. 
+
+  The SST uses a (DES) model layered on top of MPI.  The simulations are
```
Comment (gvoskuilen) [inline Tutorials/Level1Tutorial.md]: MPI & threads

File: Tutorials/Level1Tutorial.md [hunk 3]
```diff
@@ -0,0 +1,793 @@
...
+$> dot -Tpdf foo_small.dot > foo_small.pdf
+$> sst --run-mode=init --output-dot=foo_large.dot --dot-verbosity=10 /path/to/simulation/py
+$> dot -Tpdf foo_large.dot > foo_large.pdf
+```
+
+* _Converting Between Simulation File Types_: One of the latest features 
+provided by SST is the ability to develop simulation inputs using Python 
+or JSON syntax.  Python is often useful when developing large simulations 
+that require redundant components generated (multiple CPUs with 
+memory hierarchys).  However, JSON inputs are often
```
Comment (gvoskuilen) [inline Tutorials/Level1Tutorial.md]: "hierarchies"

File: Tutorials/Level1Tutorial.md [hunk 4]
```diff
@@ -0,0 +1,793 @@
...
+$> dot -Tpdf foo_large.dot > foo_large.pdf
+```
+
+* _Converting Between Simulation File Types_: One of the latest features 
+provided by SST is the ability to develop simulation inputs using Python 
+or JSON syntax.  Python is often useful when developing large simulations 
+that require redundant components generated (multiple CPUs with 
+memory hierarchys).  However, JSON inputs are often 
+easier to discern the respective graph 
+connectivity given the inherent syntax of JSON.  SST provides an internal
```
Comment (gvoskuilen) [inline Tutorials/Level1Tutorial.md]: should this be "inherent hierarchical syntax"?

File: Tutorials/Level1Tutorial.md [hunk 5]
```diff
@@ -0,0 +1,793 @@
...
+and/or contain nested sets of subcomponent configuration options.  In order 
+to assist users to discern the required or optional configuration parameters 
+for components or subcomponents, SST provides the `sst-info` command.  This 
+command outputs the relevant component and subcomponent parameters, options 
+and statistics.  There are a number of potential methods for querying 
+the configured components.  We details these as follows:
+
+* _Querying All Components_: The first method is to simply query all known 
+configured components.  By simply executing the `sst-info` command, 
+the SST-Core will output all the configuration parameters, statitics and nested
```
Comment (gvoskuilen) [inline Tutorials/Level1Tutorial.md]: "statistics"

File: Tutorials/Level1Tutorial.md [hunk 6]
```diff
@@ -0,0 +1,793 @@
...
+
+* _Executing a Simulation_: The most common command line tool is the `sst` command.  
+This is utilized to build and execute a simulation run using the SST Core and associated 
+Elements libraries.  Using the `sst` command line tool and passing a loadable 
+simulation graph using Python or JSON can be performed as follows:
+```
+$> sst /path/to/simulation.py
+$> sst /path/to/simulation.json
+```
+
```
Comment (gvoskuilen) [inline Tutorials/Level1Tutorial.md:185]: The other common basic parameter is the `--model-options="args to pass to simulation.py"` or equivalent syntax `sst /path/to/simulation.py -- <args to pass to simulation.py>`

File: Tutorials/Level1Tutorial.md [hunk 7]
```diff
@@ -0,0 +1,793 @@
...
+in the Level 2 Tutorial.
+
+Each Python-based simulation input script has requires at least three 
+key sections as outlined below:
+* **Python imports**: Each simulation input file must import the `sst` Python
+library.  This provides the fundamental component configuration options 
+and functions that link directly to the SST-Core and any configured elements.
+* **SST Core Options**: The core options section describes the top-level configuration 
+options for the SST-Core.  This includes items such as the SST program options 
+that set the baseline clock candence and simulation runtime.
```
Comment (gvoskuilen) [inline Tutorials/Level1Tutorial.md]: "cadence"

File: Tutorials/Level1Tutorial.md [hunk 8]
```diff
@@ -0,0 +1,793 @@
...
+
+The next section includes the SST Core Options.  These options drive the 
+top-level simulation configuration that are utilized across components.  Each 
+of the core program options are set with the Python function 
+`sst.setProgramOptions("Option", "Value)`.  The two most commonly 
+used program options are the `timebase` and `stop-at` options.  The 
+`timebase` option sets the baseline global clock cadence by which 
+SST will govern the core simulation components.  Each component has the ability 
+to set higher or lower fidelity clocks governing local event timing, but the 
+global clock drives the top-level simulation cadence.  Further, the `stop-at`
```
Comment (gvoskuilen) [inline Tutorials/Level1Tutorial.md]: This is not quite right. SST is DES, it doesn't have a global clock or any "core components" that are governed by a clock. However, it does have a global timestamp that is used to keep track of the current time (it jumps from time to time according to the events as you'd expect in DES), and setting timebase determines the units of that global timestamp.
Setting a higher fidelity clock than the global timebase will yield a fatal error in SST as the timestamp can't accurately represent any elapsed time interval less than the timebase. 

By default the timebase is 1ps. There are a few uncommon cases in which you'd want to set the parameter manually: 
(1) a component needs a clock with a smaller time period (in which case, set lower)
(2) you need to simulate a time period that is greater than ~2/3 of a year (max value of 1ps in a uint64_t). In this case, increase the timebase but note that any clocks will be constrained to the new timebase at minimum.

File: Tutorials/Level1Tutorial.md [hunk 9]
```diff
@@ -0,0 +1,793 @@
...
+| `stop-at`           | Simulation termination point | Simulated time (`ms`, `s`, etc) |
+| `debug-file`        | Specifies the file for debug outptu | Filename |
+| `partitioner`       | Select the graph partitioner to use | `lib.partitionerName` |
+| `verbose`           | Print information about the core runtime | Boolean (`True`, `False`)|
+| `dump_partition`    | Output the SST paritioned graph info | Boolean (`True`, `False`) |
+| `dump_config_graph` | Output the SST config graph | Boolean (`True`, `False`) |
+| `model_options`     | Provide options to the Python script | String |
+| `run-mode`          | Set the run mode | `init`, `run` or `both` |
+
+The final required section in the SST simulation script involves the definition
```
Comment (gvoskuilen) [inline Tutorials/Level1Tutorial.md:442]: The program options are not required, but this implies there are three required sections, not two (import sst and component config). Can you reword?

File: Tutorials/Level1Tutorial.md [hunk 10]
```diff
@@ -0,0 +1,793 @@
...
+* **Statistic Output** : Setting the statistic output information initializes 
+the target data output medium after the simulation has completed.  
+
+The first step in utilizing the statistics requires that the user specify 
+the statistic *load level*.  The load level specifies different levels of 
+statistics to be enabled for data collection within the target element objects.  
+Generally speaking, the higher the level, the more data will be generated.  Setting 
+the level to "0" will disable the statistics (which is also the default action).  Setting 
+the statistic load level is a simple processs of adding a single sst function call 
+to the Python input script as follows.  This is generally added before creating
```
Comment (gvoskuilen) [inline Tutorials/Level1Tutorial.md]: The order doesn't matter.

File: Tutorials/Level1Tutorial.md [hunk 11]
```diff
@@ -0,0 +1,793 @@
...
+Once you have enabled the statistics within the simulation script, you may 
+now configure the output options for the generated statistics.  By default, 
+the statistics will be output to the console.  This can be explicitly 
+set using the `sst.statOutputConsole()` function in the simulation script. However, 
+users often prefer to output the data to a useful text file that can be parsed 
+for downstream analysis.  The commonly utilized output file format is *csv*, or 
+comma separated value files.  The SST python interfaces provide a set of functions
+to enable the CSV (or other output formats) and set the required parameters.  Each 
+output enabler function is formatted as `sst.statOutputFORMAT` where `FORMAT` 
+is one of `CONSOLE`, `TXT`, `CSV`, `TXTgz` and `CXVgz`.  Each of these functions
```
Comment (gvoskuilen) [inline Tutorials/Level1Tutorial.md]: CSVgz (not CXVgz)

File: Tutorials/Level1Tutorial.md [hunk 12]
```diff
@@ -0,0 +1,793 @@
...
+for downstream analysis.  The commonly utilized output file format is *csv*, or 
+comma separated value files.  The SST python interfaces provide a set of functions
+to enable the CSV (or other output formats) and set the required parameters.  Each 
+output enabler function is formatted as `sst.statOutputFORMAT` where `FORMAT` 
+is one of `CONSOLE`, `TXT`, `CSV`, `TXTgz` and `CXVgz`.  Each of these functions 
+can take one or more parameters that define things such as the file path, 
+the output header information and the target MPI rank for which to output 
+the data from.  In our example below, we set the output to utilize *csv* 
+files, the file path to `TestOutput.csv` and the CSV separator character to 
+a comma.  As with the componenet parameters, statistics parameters can be
```
Comment (gvoskuilen) [inline Tutorials/Level1Tutorial.md]: "component"

File: Tutorials/Level1Tutorial.md [hunk 13]
```diff
@@ -0,0 +1,793 @@
...
+be called *after* the component objects are actually created.  In this manner, 
+the statistics functions are *applied* to one or more of the created component 
+objects.  Each of the statistics that reside within each component may have 
+optional parameters enabled to in order to enable/disable or modify how/when 
+the statistics are generated.  However, all statistics have a default set of 
+parameters that are enabled across all components.  These include the following:
+
+|  **Parameter**  | **Description** | **Default Value** |
+|:-|:-|:-|
+| `rate`   | Identifies the output rate of the component | 0 |
```
Comment (gvoskuilen) [inline Tutorials/Level1Tutorial.md]: Add: " 0 (output once at end of simulation)"

File: Tutorials/Level2Tutorial.md [hunk 14]
```diff
@@ -0,0 +1,1081 @@
...
+3. [Advanced JSON Scripting](#AdvJSONScripting)
+4. [Loading Parallel Graphs](#LoadingParallelGraphs)
+5. [Links to External SST Components](#ExternSSTComp)
+
+## Parallel SST runs: <a name="ParallelSSTRuns"></a>
+
+### Introduction and Environment
+
+One of the key features of SST is the ability to execute parallel 
+simulations that utilize the MPI (Message Passing Interface) to span
```
Comment (gvoskuilen) [inline Tutorials/Level2Tutorial.md]: add: and/or C++ threads

**Review by gvoskuilen — CHANGES_REQUESTED** [review]

File: Tutorials/Level2Tutorial.md [hunk 15]
```diff
@@ -0,0 +1,1081 @@
...
+are natively built to utilize MPI in order to perform communication between 
+individual components across multiple MPI "ranks."  This enables users to 
+build and execute large scale simulations without fear of exhausting a single 
+system's resources.  When a user requests a parallel simulation (using MPI 
+or threading), the SST Core infrastructure will automatically partition the 
+requested components and subcomponents across the target number of parallel 
+units.  This partitioning considers the requested simulation components, the 
+connectivity of the components and the latency of the links between components 
+in order to co-locate components and subcomponents that are expected to frequently 
+exchange events.
```
Comment (gvoskuilen) [inline Tutorials/Level2Tutorial.md]: This is a bit of an overstatement. Maybe change to "in order to minimize synchronization overhead".

Some background: SST has a few built-in automatic partitioning options (run `sst-info sst` to view), in addition to accepting a manual partitioning. None of the automatic options take into account how frequently events are expected to be exchanged. That would be a good metric, as would co-locating components that interact more frequently, but the core doesn't have enough info to predict this. The `simple` partitioner does take into account the graph structure as it tries to cut high-latency links and spread components evenly across ranks/threads. The goal is to minimize synchronization frequency (cutting high latency links) while balancing work across ranks (number of components is a _very_ loose proxy for amount of work, but one that is easy to compute ahead of time).

File: Tutorials/Level2Tutorial.md [hunk 16]
```diff
@@ -0,0 +1,1081 @@
...
+
+### Introduction
+
+The Python SST model interfaces provide users a wide array of options 
+for constructing scalable simulations using composable interfaces.  The benefit 
+to using Python as the basis for the simulation input is that users may utilize 
+all the standard Python loop constructs, flow control and function interfaces. The 
+standard Python syntax combined with the SST Python classes enable powerful 
+features for creating simulations.  The remainder of this section provides an overview 
+into a number of advanced Python features as well as the SST Python interfaces.  THis
```
Comment (gvoskuilen) [inline Tutorials/Level2Tutorial.md]: Minor fix: "THis" -> "This"

File: Tutorials/Level2Tutorial.md [hunk 17]
```diff
@@ -0,0 +1,1081 @@
...
+into a number of advanced Python features as well as the SST Python interfaces.  THis 
+section assumes that the user has basic knowledge of Python syntax and data 
+structures.
+
+### Program Options
+
+The SST Python interfaces provide users the ability to to retrieve and set global 
+program options.  These options allow users to control global options that 
+are analogous to those passed on the command line.  This is often useful when 
+buidling large, complex simulations that need to be utilized across multiple users
```
Comment (gvoskuilen) [inline Tutorials/Level2Tutorial.md]: "buidling" -> "building"

File: Tutorials/Level2Tutorial.md [hunk 18]
```diff
@@ -0,0 +1,1081 @@
...
+Opts = sst.getProgramOptions()
+
+if Opts["verbose"] == 1 :
+  # print all the program options
+  print(Opts)
+```
+
+Similarly, we can enable certain global program options using the 
+`sst.setProgramOption` interface.  This interface accepts both single name:value 
+pairs as well as full dictionaries of name:value pairs.  In this example,
```
Comment (gvoskuilen) [inline Tutorials/Level2Tutorial.md]: `sst.setProgramOption` takes two parameters, a name and a value (both strings)
`sst.setProgramOptions` takes a dictionary of name:value pairs

File: Tutorials/Level2Tutorial.md [hunk 19]
```diff
@@ -0,0 +1,1081 @@
...
+### Building Complex Simulations
+
+Now that we have the ability to query the environment, set 
+program set and retrieve command line arguments via the SST Python 
+environment, we can start building more complex simulation scripts 
+using all of the aforementioned features coupled to the SST Python 
+function interfaces.
+
+Before we begin our next example, we need to 
+introduct several new SST Python functions, `pushNamePrefix` and
```
Comment (gvoskuilen) [inline Tutorials/Level2Tutorial.md]: "introduct" -> "introduce"

File: Tutorials/Level2Tutorial.md [hunk 20]
```diff
@@ -0,0 +1,1081 @@
...
+using all of the aforementioned features coupled to the SST Python 
+function interfaces.
+
+Before we begin our next example, we need to 
+introduct several new SST Python functions, `pushNamePrefix` and 
+`popNamePrefix`.  As mentioned in the Level 1 tutorial material, 
+each component must retain a unique name internal to SST.  This ensures 
+that subcomponents and links are attached to the correct components in the 
+prescribed manner.  When constructing simulation scripts that utilize 
+a large number of similar components, its often difficult to manually
```
Comment (gvoskuilen) [inline Tutorials/Level2Tutorial.md]: "its" -> "it's" or "it is"

File: Tutorials/Level2Tutorial.md [hunk 21]
```diff
@@ -0,0 +1,1081 @@
...
+  CLOCK #3 - TICK Num 12; Param = 333
+  CLOCK #3 - TICK Num 13; Param = 333
+  CLOCK #3 - TICK Num 14; Param = 333
+  CLOCK #3 - TICK Num 15; Param = 333
+Simulation is complete, simulated time: 100 s
+```
+
+If we increase the number of generated clocks on the command line, 
+we see the following.  Further, if we enable verbosity, we also 
+find that SST echoes a valid Python script back out to config.py
```
Comment (gvoskuilen) [inline Tutorials/Level2Tutorial.md:432]: Maybe missed it, are you actually increasing the verbosity in the example? Both invocations (above/below) pass '-v' but they have different input files. Have you shown these two files somewhere (I see one)? Also, -v will not echo the python script back out, that's the `--output-config=FILE` option.
 
I'm looking at the invocation above:
`sst -v --model-options="--clocks 1" sim.py`
and below:
`sst -v --model-options="--clocks 2" test2.py`

File: Tutorials/Level2Tutorial.md [hunk 22]
```diff
@@ -0,0 +1,1081 @@
...
+})
+comp_DOMAIN1_clocker.setCoordinates(0, 0, 0)
+# Define SST Statistics Options:
+sst.setStatisticOutput("sst.statOutputConsole")
+# End of generated output.
+```
+
+### Python Classes
+
+The following section projects an outline of the available SST Python
```
Comment (gvoskuilen) [inline Tutorials/Level2Tutorial.md:565]: Can you add a note that these are accurate as of SST 11.1?

File: Tutorials/Level2Tutorial.md [hunk 23]
```diff
@@ -0,0 +1,1081 @@
...
+The following are several candidate examples of parallel JSON input files using 
+standard SST tests.
+
+|  **Description**  | **Ranks** | **Rank Files** |
+|:-|:-:|:-|
+| Miranda Stencil3D | 2 | [Rank0](samples/parallel/stencil3dbench0.json) [Rank1](samples/parallel/stencil3dbench1.json)|
+| Merlin 256 Node FatTree | 4 | [Rank0](samples/parallel/fattree_256_test_parallel0.json) [Rank1](samples/parallel/fattree_256_test_parallel1.json) [Rank2](samples/parallel/fattree_256_test_parallel2.json) [Rank3](samples/parallel/fattree_256_test_parallel3.json)|
+
+## Links to External SST Components: <a name="ExternSSTComp"></a>
+* [HBM DRAMSIm2](https://github.com/tactcomplabs/HBM/releases/tag/sst-8.0.0-release)
```
Comment (gvoskuilen) [inline Tutorials/Level2Tutorial.md]: Is this section just a link to what's available? Can you instead point users to check: http://sst-simulator.org/SSTPages/SSTMainDocumentation/ and click on "External Components Instructions" (the actual page URL has is linked to a specific release)? This has a complete list as well as links to the versions/tags we test against. There's also a page that links to community elements that we don't necessarily test against. (http://sst-simulator.org/SSTPages/SSTMainCommunity/)

File: Tutorials/Level1Tutorial.md (hunk 1 above)
Comment (jleidel) [reply Tutorials/Level1Tutorial.md]: revised in branch

File: Tutorials/Level1Tutorial.md (hunk 2 above)
Comment (jleidel) [reply Tutorials/Level1Tutorial.md]: resolved in branch

File: Tutorials/Level1Tutorial.md (hunk 3 above)
Comment (jleidel) [reply Tutorials/Level1Tutorial.md]: resolved in branch

File: Tutorials/Level1Tutorial.md (hunk 4 above)
Comment (jleidel) [reply Tutorials/Level1Tutorial.md]: resolved in branch

File: Tutorials/Level1Tutorial.md (hunk 5 above)
Comment (jleidel) [reply Tutorials/Level1Tutorial.md]: resolved in branch

File: Tutorials/Level1Tutorial.md (hunk 7 above)
Comment (jleidel) [reply Tutorials/Level1Tutorial.md]: resolved in branch

File: Tutorials/Level1Tutorial.md (hunk 6 above)
Comment (jleidel) [reply Tutorials/Level1Tutorial.md:185]: added to branch

File: Tutorials/Level1Tutorial.md (hunk 11 above)
Comment (jleidel) [reply Tutorials/Level1Tutorial.md]: revised in branch

File: Tutorials/Level1Tutorial.md (hunk 12 above)
Comment (jleidel) [reply Tutorials/Level1Tutorial.md]: resolved in branch

File: Tutorials/Level1Tutorial.md (hunk 13 above)
Comment (jleidel) [reply Tutorials/Level1Tutorial.md]: added to branch

File: Tutorials/Level1Tutorial.md (hunk 10 above)
Comment (jleidel) [reply Tutorials/Level1Tutorial.md]: removed unnecessary statement

File: Tutorials/Level1Tutorial.md (hunk 8 above)
Comment (jleidel) [reply Tutorials/Level1Tutorial.md]: reworded this block; noted that timebase is option and 1ps is the default

File: Tutorials/Level1Tutorial.md (hunk 9 above)
Comment (jleidel) [reply Tutorials/Level1Tutorial.md:442]: reworded and noted that program options are optional; now two required sections

File: Tutorials/Level2Tutorial.md (hunk 14 above)
Comment (jleidel) [reply Tutorials/Level2Tutorial.md]: updated in the branch

File: Tutorials/Level2Tutorial.md (hunk 16 above)
Comment (jleidel) [reply Tutorials/Level2Tutorial.md]: changed in branch

File: Tutorials/Level2Tutorial.md (hunk 17 above)
Comment (jleidel) [reply Tutorials/Level2Tutorial.md]: changed in branch

File: Tutorials/Level2Tutorial.md (hunk 19 above)
Comment (jleidel) [reply Tutorials/Level2Tutorial.md]: changed in the branch

File: Tutorials/Level2Tutorial.md (hunk 20 above)
Comment (jleidel) [reply Tutorials/Level2Tutorial.md]: changed in the branch

File: Tutorials/Level2Tutorial.md (hunk 18 above)
Comment (jleidel) [reply Tutorials/Level2Tutorial.md]: fixed in the branch

File: Tutorials/Level2Tutorial.md (hunk 22 above)
Comment (jleidel) [reply Tutorials/Level2Tutorial.md:565]: duly noted.  Added this to the text

File: Tutorials/Level2Tutorial.md (hunk 15 above)
Comment (jleidel) [reply Tutorials/Level2Tutorial.md]: Duly noted.  This is an area where we are very interested to help in the future for very large graphs.  I made the change in the documentation though.

File: Tutorials/Level2Tutorial.md (hunk 21 above)
Comment (jleidel) [reply Tutorials/Level2Tutorial.md:432]: I believe this was a bug in copying/pasting from the tests we generated in our local environment.  Fixed in our local branch

File: Tutorials/Level2Tutorial.md (hunk 23 above)
Comment (jleidel) [reply Tutorials/Level2Tutorial.md]: What?  No selfish plug?  :-)
Fixed in the branch

**Pull request merged**